### DISCUSSION TITLE
Stop running migrations by default

### DISCUSSION THREAD

**Discussion by abhjitsantra** [elsa-workflows/elsa-core#4438] [Q&A] [answered]
I am using Elsa v3 and I have created a separate project to deal with the schema migrations. The plan is to run the migrations only when I upgrade the Elsa packages. The problem I am having is that, though I have created fresh migration and applied on the database after package upgrade, the project complains about some table already exists during startup. How can I deal with this? I can't afford deleting the DB each time I do upgrade Elsa packages.

**Comment by sfmskywalker** [comment, marked as answer]
Hello @abhjitsantra 

I completely understand, and maintaining your own migrations is the recommended approach, because this gives you full control over when and how you update your schemas.

So, you are right to want to opt-out of automatically running migrations. To do so, set the `RunMigrations` flag to `false` when configuring a feature with EF Core.

For example:

```csharp
services.AddElsa(elsa => 
{
   elsa.UseIdentity(identity => identity.UseEntityFrameworkCore(ef => ef.RunMigrations = false));
   elsa.UseManagement(management => management.UseEntityFrameworkCore(ef => ef.RunMigrations = false));
   elsa.UseRuntime(runtime => runtime.UseEntityFrameworkCore(ef => ef.RunMigrations = false));
})
```

I hope this helps!

**Reply by abhjitsantra** [reply]
Thanks @sfmskywalker.
I'll explore more to see if there is any impact on the already saved workflows, specifically with the migrations where there is a data loss.

**Reply by jsUnitedLex** [reply]
@sfmskywalker I'm coming up short running the migrations manually in our pipelines using an EF bundle. Is this something that is supported?

**Reply by sfmskywalker** [reply]
@jsUnitedLex Yes that's supported. Although I haven't confirmed the exact steps, here are some pointers you might explore:

- Disable automatically running migrations (see my previous message explaining how to do that)
- Add a pipeline step that uses the dotnet cli EF Core tool to apply the necessary migrations.

That should be it.

**Reply by jsUnitedLex** [reply]
Thank you

**Reply by jsUnitedLex** [reply]
@sfmskywalker Is this path still valid as well?
https://v2.elsaworkflows.io/docs/installation/persistence#running-migrations-manually
If so, should we be using the v2 persistence packages or are there others for v3?

**Reply by sfmskywalker** [reply]
Although that document targets running Elsa 2 migrations, the general principle is the same for Elsa 3 in case you wih to generate and maintain your own migrations:

- Create your own design-time DB context factory classes for the various DB contexts you want to generate migrations for (take a look at the elsa-extensions project in the main branch and look for the Elsa.Persistence.EFCore. In there, you'll find a Modules folder containing DB context classes for which you will want to generate migrations.
- Run the appropriate CLI command to generate the migrations.

In case you only want to update the database with existing migrations, you only need to use the appropriate CLI command for that and specify the existing DB context provided by the Elsa.Persistence.EFCore.* packages (in case of 3.6.0-preview.xxx - we have **renamed** the packages from Elsa.EntityFrameworkCore,* to Elsa.Persistence.EFCore.*).

In any case, don't use any **v2** packages when using Elsa 3 packages.

**Comment by sfmskywalker** [comment]
The migration-control guidance above is still the applicable approach: set `RunMigrations = false` for each EF Core persistence feature used by the application, then run the matching Elsa 3 EF Core contexts from a controlled deployment/pipeline step. Do not use Elsa 2 persistence packages with Elsa 3.

The current EF Core persistence feature records migration behavior per `DbContext` in `MigrationOptions`: https://github.com/elsa-workflows/elsa-core/blob/main/src/modules/Elsa.Persistence.EFCore.Common/PersistenceFeatureBase.cs#L42-L67.

Closing this resolved configuration thread. If a current Elsa 3 migration bundle fails, please open a new discussion with the Elsa version, provider, context name, and the complete command/output.
